### PR TITLE
Handling of EmailReference messages

### PR DESCRIPTION
**Pull request by moodmosaic**
This pull request handles the EmailReference pointer message (related to #4) as discussed in #18.

It has been tested with:
- [Mailtrap.io](https://mailtrap.io/)<sup>*</sup> for attachments of small size ~1MB
- the technique I previously described [here](https://github.com/GreanTech/Qaiain/pull/18/files#r14179272) for attachments of size ~20MB

---

<sup>*</sup>Mailtrap.io rocks!, by the way. :)

### CODE REVIEW

File: Src/Qaiain.CommandLine/Program.fs [hunk 1]
```diff
@@ -107,6 +108,12 @@ module Mail =
     let private toMailAddress address =
         MailAddress(address.SmtpAddress, address.DisplayName)
 
+    let private toAttachment attachment =
+        new Attachment(
```
Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs:112]: Disposed when the MailMessage is disposed [(source)](http://referencesource.microsoft.com/#System/net/System/Net/mail/MailMessage.cs#280).

File: Src/Qaiain.CommandLine/Program.fs [hunk 2]
```diff
@@ -130,6 +140,22 @@ let queue =
     q.CreateIfNotExists() |> ignore
     q
 
+let blob =
```
Comment (ploeh) [inline Src/Qaiain.CommandLine/Program.fs]: Isn't this a public value? If so, is `blob` a good, descriptive name?

File: Src/Qaiain.CommandLine/App.config [hunk 3]
```diff
@@ -3,6 +3,7 @@
   <appSettings>
     <add key="storageConnectionString" value="" />
     <add key="queue-name" value="qaiain"/>
+    <add key="blob-name" value="qaiain"/>
```
Comment (ploeh) [inline Src/Qaiain.CommandLine/App.config]: Isn't this a name of a BLOB container?

File: Src/Qaiain.CommandLine/Program.fs [hunk 4]
```diff
@@ -130,6 +140,22 @@ let queue =
...
+    let storageAccount =
+        CloudConfigurationManager.GetSetting "storageConnectionString"
+        |> CloudStorageAccount.Parse
+
+    let name = CloudConfigurationManager.GetSetting "blob-name"
+    let blob = storageAccount.CreateCloudBlobClient().GetContainerReference(name)
+    blob.CreateIfNotExists() |> ignore
+    blob
+
+let download (b : Blob.CloudBlockBlob) =
```
Comment (ploeh) [inline Src/Qaiain.CommandLine/Program.fs]: What's wrong with the [DownloadText](http://msdn.microsoft.com/en-us/library/microsoft.windowsazure.storage.blob.cloudblockblob.downloadtext.aspx) method?

File: Src/Qaiain.CommandLine/Program.fs [hunk 5]
```diff
@@ -144,15 +170,24 @@ let send =
...
+let rec handle msg =
+    match msg |> Mail.parse with
+    | Mail.EmailData mail ->
+        send mail
+    | Mail.EmailReference ref ->
+        let b = blob.GetBlockBlobReference(ref.DataAddress)
+        b
+        |> download
+        |> handle
+        b.Delete()
```
Comment (ploeh) [inline Src/Qaiain.CommandLine/Program.fs]: What happens if this delete operation succeeds, but then the subsequent `queue.DeleteMessage msg` operation fails?

File: Src/Qaiain.CommandLine/App.config (hunk 3 above)
Comment (moodmosaic) [inline Src/Qaiain.CommandLine/App.config]: Yes. Perhaps `blob-container-name` then. Or even `container-name`. What do you think?

File: Src/Qaiain.CommandLine/Program.fs (hunk 2 above)
Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: How about `blobContainer` or even `container`?

File: Src/Qaiain.CommandLine/Program.fs (hunk 5 above)
Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: We might end up with orphaned messages in the queue, probably. How can we prevent this?

File: Src/Qaiain.CommandLine/Program.fs (hunk 4 above)
Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: Indeed.

File: Src/Qaiain.CommandLine/App.config (hunk 3 above)
Comment (ploeh) [inline Src/Qaiain.CommandLine/App.config]: I think "messages-container" would suffice.

File: Src/Qaiain.CommandLine/Program.fs (hunk 2 above)
Comment (ploeh) [inline Src/Qaiain.CommandLine/Program.fs]: I don't think either of those names are particularly descriptive of the _role_ that particular container plays...

An alternative to a rename might also be to make the value private...

File: Src/Qaiain.CommandLine/Program.fs (hunk 5 above)
Comment (ploeh) [inline Src/Qaiain.CommandLine/Program.fs]: Yes, and what happens if the message is left in the queue?

Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: Eventually it expires - until then we will probably end up processing it (even multiple times) for no reason...

Comment (ploeh) [inline Src/Qaiain.CommandLine/Program.fs]: What happens the next time the message is picked up from the queue?

Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: So... if we get a StorageException with message _The remote server returned an error: (404) Not Found_ we can probably delete the orphaned message from the queue.

File: Src/Qaiain.CommandLine/Program.fs (hunk 2 above)
Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: It is now private in 3c9f212.

File: Src/Qaiain.CommandLine/App.config (hunk 3 above)
Comment (moodmosaic) [inline Src/Qaiain.CommandLine/App.config]: Done in 3aa6728.

File: Src/Qaiain.CommandLine/Program.fs (hunk 4 above)
Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: Done in 14aa2e2.

File: Src/Qaiain.CommandLine/App.config [hunk 6]
```diff
@@ -3,6 +3,7 @@
   <appSettings>
     <add key="storageConnectionString" value="" />
     <add key="queue-name" value="qaiain"/>
+    <add key="messages-container" value="qaiain"/>
```
Comment (ploeh) [inline Src/Qaiain.CommandLine/App.config]: Could we please make the default value "messages", or something like that? Hypothetically, it's only one container among several; it's like a file folder...

Comment (moodmosaic) [inline Src/Qaiain.CommandLine/App.config]: Hm... If we have "messages-container" on the left and "messages" on the right, isn't it a bit confusing? How about "large-messages"? No?

File: Src/Qaiain.CommandLine/Program.fs (hunk 5 above)
Comment (ploeh) [inline Src/Qaiain.CommandLine/Program.fs]: Yes, although I'd like to reword this a bit.

A message reference that points to a non-existent BLOB indicates that the message was already handled. Handling the message from the queue should be an idempotent operation, so I agree that the message should be deleted from the queue.

However, I _would_ prefer that we hide the implementation details of StorageException in a function, and get a bit of test coverage around this behaviour.

Would it make sense to add a new case to the `EmailMessage` Discriminated Union? A case that captures the state where the email reference points to a BLOB that's no longer there. It's not quite `Unknown`, but rather something like `OrphanedEmailReference`.

File: Src/Qaiain.CommandLine/App.config (hunk 6 above)
Comment (ploeh) [inline Src/Qaiain.CommandLine/App.config]: _messages-container_ is the role of the container; _messages_ would be its default name.

It's like having a file directory called _messages_, and a configuration value called _messages-directory_. Is that confusing?

Comment (moodmosaic) [inline Src/Qaiain.CommandLine/App.config]: I think it's not. It's now renamed to _messages_ in e804112.

File: Src/Qaiain.CommandLine/Program.fs (hunk 5 above)
Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: The `EmailMessage` Discriminated Union represents the content of the parsed XML.

A hypothetical `OrphanedEmailReference` doesn't represents the content of the parsed XML. It looks more like a storage related concept to me.

AFAICT the `parse` function can't return an `OrphanedEmailReference` unless we attempt to download the blob's contents.

A better place would probably be the new `handle` function. Its signature though is:

```
string -> unit
```

In other words I am not sure
- what is the best place to capture and return an `OrphanedEmailReference`
- how to get a decent code coverage around it

Comment (ploeh) [inline Src/Qaiain.CommandLine/Program.fs]: You are probably right that the correct place to handle this is in the `handle` function itself, but I'd like to get that function under test, then.

As given at the moment, the `handle` function closes over the `blob` value and the `send` function. If, instead of closing over them, it took them as parameters, it would be almost testable. If we also change the input so that instead of passing a `CloudBlobContainer`, it would pass another function that returns a `string`, it would be testable.

Something like this:

``` F#
let rec handle (getMessage : string -> string) (send : EmailData -> unit) msg =
```

(Not that I'm suggesting that we should have explicit type annotations on the function - the above is just to illustrate what I mean...)

Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: I see, and so that way we probably don't necessarily need the `OrphanedEmailReference`, yes?

Comment (ploeh) [inline Src/Qaiain.CommandLine/Program.fs]: Right, `OrphanedEmailReference` was a bad idea.

Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: I think, now we also need to pass a function as parameter that deletes the BLOB since `blob` is not accessed inside the `handle` function anymore.

Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: Something like this:

```
let rec handle (getMessage) (deleteMessage) (sendEmail) msg =
```

Comment (ploeh) [inline Src/Qaiain.CommandLine/Program.fs]: Let's start with that - it's a bit too many arguments to my taste, but it's a decent starting point, so let's start with that, and see what we can learn from this direction...

Just thinking out loud... Such a `getMessage`/`deleteMessage` pair almost sounds like some sort of context in itself. In C#, many people would model it with a `using` block. In F#, perhaps it could be modelled as a monadic type...

However, let's not go there right now... let's get it working first - it's not _that_ complicated, after all.

Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: Agreed about the number of arguments - they are more than 3... So, let's start with that and hopefully soon we can make these functions composable with a monad.

Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: I think, we can move the `handle` function inside the `Mail` module. Do you agree?

image

Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: FWIW, I made an attempt to move it and then some compiler errors appeared. So, perhaps we can consider moving it on a separate pull request, then.

Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: I just added [some commits](https://github.com/moodmosaic/Qaiain/compare/e804112f531b4ff486e0338cdd9afa646804fb81...f0477c74efefb8b5ce9f1e50ef75094e9274ea06) to get the `handle` function under test.

File: Src/Qaiain.UnitTests/MailTests.fs [hunk 7]
```diff
@@ -307,3 +307,161 @@ let ParseReturnsCorrectResult () =
     |> Seq.map (fun tc -> TestCase (fun () ->
         let actual = parse tc.input
         verify <@ tc.expected = actual @>))
+
+open Program
```
Comment (moodmosaic) [inline Src/Qaiain.UnitTests/MailTests.fs:311]: This prompts me to consider moving the `handle` function to the `Mail` module.

File: Src/Qaiain.CommandLine/Program.fs (hunk 5 above)
Comment (ploeh) [inline Src/Qaiain.CommandLine/Program.fs]: Moving the `handle` function to the `Mail` module sounds like a good idea, but _you're_ introducing the function in this Pull Request (PR), so you're allowed to have a stronger opinion of its design than me :)

Still, if it's difficult to move, I'll also be happy dealing with this in a separate PR.

**Comment by ploeh** [comment]
Thank you for your contribution!

**Comment by moodmosaic** [comment]
Thank you for helping!

File: Src/Qaiain.CommandLine/Program.fs (hunk 5 above)
Comment (moodmosaic) [inline Src/Qaiain.CommandLine/Program.fs]: I just [opened](https://github.com/GreanTech/Qaiain/pull/24) a separate Pull Request (PR) (since this PR is merged) :)
